### ISSUE TITLE
About 'dp_1 = self.sigmoid(dp_1) * 0.3'

### ISSUE THREAD

**Issue by daydreamer2023** [visinf/multi-mono-sf#5]
Hi!
Thanks for making your code open.
Could you please further explain why the parameter chooses 0.3?
Why not `dp_1 = self.sigmoid(dp_1)`?

https://github.com/visinf/multi-mono-sf/blob/8e71381635c7ad4981055301705e018a884a6085/models/model_monosceneflow.py#L122

**Comment by hurjunhwa** [comment]
Hi, 
We adopt this approach from [Monodepth](https://github.com/mrharicot/monodepth).

It scales the disparity output between [0, 0.3] for better optimization.
They assume that the range of disparity value is between [0, 0.3] x (image width).

https://github.com/mrharicot/monodepth/blob/b76bee4bd12610b482163871b7ff93e931cb5331/monodepth_model.py#L123

**Comment by daydreamer2023** [comment]
Thank you so much for your quick reply！

**Comment by daydreamer2023** [comment]
Hi!
could you please provide the `logbbok.txt` file?
I want to observe the loss change of the following training process, but its training process is very time-consuming. It takes me 1.5 hours to run an epoch.

**Comment by hurjunhwa** [comment]
Hi,

Here is a tensorboard file if it can help.
https://drive.google.com/file/d/1nX1WOff-tUd5wni4N3-3pxyKIZgBvkBY/view?usp=sharing

**Comment by daydreamer2023** [comment]
Hi,
Why your evaluation loss is far smaller than training loss? (The yellow one is your loss function curve.)
Thanks a lot.

![image](https://user-images.githubusercontent.com/19202609/139877787-7bc9da1f-8ab9-4fa5-b045-a4b0ea4c1c1a.png)

**Comment by hurjunhwa** [comment]
For that experiment, I think the numbers for "Val" were the scene flow accuracy (between 0 and 1), not the validation loss.

**Comment by daydreamer2023** [comment]
That sounds reasonable.
Thanks a lot!

**Comment by daydreamer2023** [comment]
Hi,
I want to ask you two questions.
**1st:**
https://github.com/visinf/multi-mono-sf/blob/8e71381635c7ad4981055301705e018a884a6085/models/modules_sceneflow.py#L235-L236

About `-20.0 * (0.4-disp)`, are they magic numbers determined based on experience or derive from theory ?

**2rd:**
https://github.com/visinf/multi-mono-sf/blob/8e71381635c7ad4981055301705e018a884a6085/models/modules_sceneflow.py#L219
https://github.com/visinf/multi-mono-sf/blob/8e71381635c7ad4981055301705e018a884a6085/models/modules_sceneflow.py#L237

What is the main function of designing `self.conv1x1`?

**Comment by hurjunhwa** [comment]
Hi,

It's a kind of hyper-parameters for the softmax function.
https://github.com/sniklaus/softmax-splatting/blob/7ba7a19dc054674ae0407674372b6f0d8a36b2c7/run.py#L58
-20 is a sort of temperature or weight constant for the softmax function, and (0.4-disp) is to make it inverse (given that disp ranges between [0, 0.3]. (well actually (0.3-disp) could make more sense..)

``self.conv1x1`` is to learn to scale the dot product value of two features before thresholding with 0.5.

Hopefully it helped your understanding!

**Comment by daydreamer2023** [comment]
Thank you very much, your answer helped me a lot！
Using `self.conv1x1` to to learn to scale the dot product value Is a great idea.
This should be equivalent to a learned pixel-wise threshold.

**Comment by hurjunhwa** [comment]
Right, it should be equivalent :)

**Comment by daydreamer2023** [comment]
https://github.com/visinf/multi-mono-sf/blob/8e71381635c7ad4981055301705e018a884a6085/models/model_monosceneflow.py#L110-L111
Hi, one more question.
Are `x1`,` x0`, and `x2` in your code specially designed like this?
```
x_out_f, sf_f_res, dp_f = self.flow_estimators[l]( ... , x_outs_f_pr[l], fl_pr, dp_pr, x0, x1) 
x_out_b, sf_b_res, dp_b = self.flow_estimators[l](..., x_outs_b_pr[l], fl_pr, dp_pr, x0, x1) 
```
Is the following order better?
```
x_out_f, sf_f_res, dp_f = self.flow_estimators[l]( ... , x_outs_f_pr[l], fl_pr, dp_pr, x1, x0) 
x_out_b, sf_b_res, dp_b = self.flow_estimators[l](..., x_outs_b_pr[l], fl_pr, dp_pr, x1, x2) 
```

**Comment by hurjunhwa** [comment]
Hi,
Here, ``x0`` and ``x1`` are used to forward-warp and propagate the previous hidden & cell states from the **previous time step**, through the softmax splatting.
Thus, for either forward scene flow or backward scene flow, it needs the feature map from the previous time step (i.e., ``x0``)
https://github.com/visinf/multi-mono-sf/blob/8e71381635c7ad4981055301705e018a884a6085/models/modules_sceneflow.py#L302-L303

During the soft-splatting, it compares the feature similarity between warped ``x0`` (i.e., previous time step) and ``x1`` (current time step).

https://github.com/visinf/multi-mono-sf/blob/8e71381635c7ad4981055301705e018a884a6085/models/modules_sceneflow.py#L237

Here, `feat1_warp` corresponds to `x0`, and `feat2` is actually `x1`.

**Comment by daydreamer2023** [comment]
Yes， I get it.
Thank you very much for your patient answer！！！
